plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java

```diff
@@ -18,45 +18,76 @@
     private static final Integer NUMBER_OF_REPLICA = 3;
 
+    private static final String TOXIPROXY_IMAGE = "shopify/toxiproxy:2.1.0";
+    private static final String TOXIPROXY_NETWORK_ALIAS = "toxiproxy";
+
+    private final Network network;
+    private final ToxiproxyContainer toxiProxy;
     private final GenericContainer<?> master;
     private final List<GenericContainer<?>> tServers;
 
+    /**
+     * WORKAROUND WARNING:
+     * Kudu tablets needs to know the host/mapped port it will be bound to in order to configure --rpc_advertised_addresses
+     * However when using non-fixed ports in testcontainers, we only know the mapped port after the container starts up
+     * In order to workaround this, create a proxy to forward traffic from the host to the underlying tablets
+     * Since the ToxiProxy container starts up *before* kudu, we know the mapped port when configuring the kudu tablets
+     */
     public TestingKuduServer()
     {
-        Network network = Network.newNetwork();
+        network = Network.newNetwork();
         ImmutableList.Builder<GenericContainer<?>> tServersBuilder = ImmutableList.builder();
-        this.master = new GenericContainer<>("apache/kudu:1.10.0")
+
+        String hostIP = getHostIPAddress();
+
+        String masterContainerAlias = "kudu-master";
+        this.master = new GenericContainer<>(KUDU_IMAGE)
                 .withExposedPorts(KUDU_MASTER_PORT)
                 .withCommand("master")
+                .withEnv("MASTER_ARGS", format("--fs_wal_dir=/var/lib/kudu/master --logtostderr --use_hybrid_clock=false --rpc_bind_addresses=%s:%s", masterContainerAlias, KUDU_MASTER_PORT))
                 .withNetwork(network)
-                .withNetworkAliases("kudu-master");
+                .withNetworkAliases(masterContainerAlias);
+
+        toxiProxy = new ToxiproxyContainer(TOXIPROXY_IMAGE)
+                .withNetwork(network)
+                .withNetworkAliases(TOXIPROXY_NETWORK_ALIAS);
+        toxiProxy.start();
 
         for (int instance = 0; instance < NUMBER_OF_REPLICA; instance++) {
             String instanceName = "kudu-tserver-" + instance;
-            GenericContainer<?> tableServer = new GenericContainer<>("apache/kudu:1.10.0")
+            ToxiproxyContainer.ContainerProxy proxy = toxiProxy.getProxy(instanceName, KUDU_TSERVER_PORT);
+            GenericContainer<?> tableServer = new GenericContainer<>(KUDU_IMAGE)
                     .withExposedPorts(KUDU_TSERVER_PORT)
                     .withCommand("tserver")
-                    .withEnv("KUDU_MASTERS", "kudu-master:" + KUDU_MASTER_PORT)
+                    .withEnv("KUDU_MASTERS", format("%s:%s", masterContainerAlias, KUDU_MASTER_PORT))
+                    .withEnv("TSERVER_ARGS", format("--fs_wal_dir=/var/lib/kudu/tserver --logtostderr --use_hybrid_clock=false --rpc_bind_addresses=%s:%s --rpc_advertised_addresses=%s:%s", instanceName, KUDU_TSERVER_PORT, hostIP, proxy.getProxyPort()))
```

Comment:
I think we could remove `rpc_bind_address` here

Reply:
Here I think it is better to be explicit than implicit.